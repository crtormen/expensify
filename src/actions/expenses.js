import uuid from 'uuid';
import database from '../firebase/firebase';


// Actions ate constants where are defined the arguments and their respectives default states, 
// for the object returned, where those args are applied to be executed in the reducers when dispatched by redux-store


//Action generator ADD_EXPENSE
export const addExpense = (expense) => ({ 
    type: 'ADD_EXPENSE',
    expense
});

// Pass a function instead of a object  
export const startAddExpense = (expenseData = {}) => {
    return (dispatch) => { 
        const { //default values
            description = '', 
            note = '', 
            amount = 0, 
            createdAt = 0
        } = expenseData;
        const expense = { description, note, amount, createdAt };

        return database.ref('expenses').push(expense).then((ref) => { 
            dispatch(addExpense({
                id: ref.key,
                ...expense
            }));
        });
    };
};

//Action Generator REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//Action Generator EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates
})

//Action Generator SET_EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//Action Generator START_SET_EXPENSES
export const startSetExpenses = () => {
    return (dispatch) => {
        //returns a promises, so that when use startSetExpenses, we can use "then"
        return database.ref('expenses').once('value').then((snapshot) => {
            let expenses = [];
            snapshot.forEach((childSnapshot) => {
                expenses.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                })
            });

            dispatch(setExpenses(expenses));
        });        
    };
};

