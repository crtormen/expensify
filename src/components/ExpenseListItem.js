import React from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

const ExpenseListItem = ({ id, description, amount, createdAt }) => ( //Destructuring out of the props
    <div className="list-item">
        <h3><Link to={`/edit/${id}`}>{description}</Link></h3> 
        <p>Amount: {numeral(amount/100).format('$0,0.00')} - Date: {moment(createdAt).format('DD/MM/YY')}</p>
    </div>
);


export default ExpenseListItem; //With no mapping to state, connect still provides access to dispatch