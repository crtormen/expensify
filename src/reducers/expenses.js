//Expenses Reducer

//Place to define default state, and the function that execute the actions related to expenses.

const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //return state.concat(action.expense); //Return a new array, instead of change the state, like push
            return [...state, action.expense]; //Does the same (create a new a array)
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id ); //return all in a new array, except that one where filter not match (id == action.id)
        case 'EDIT_EXPENSE':
            return state.map((expense) => {  //looks to expense list in state 
                if (expense.id === action.id) { //trying to find the expense defined in action.id
                    return { 
                        ...expense,
                        ...action.updates //Object spread operator (create new object based ond the previous, with updates and concatenations)
                    }
                }
                return expense; //do nothing if not found
            });
        case 'SET_EXPENSES':
            return action.expenses;
        default: 
            return state;
    }
};

//export default expensesReducer;