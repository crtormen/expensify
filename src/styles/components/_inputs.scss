.text-input {
    border: 1px solid #cacccd;
    height: 50px;
    font-size: $font-size-large;
    font-weight: 300;
    padding: $s-size;
}

.select {
    @extend .text-input;
    font-size: $font-size-small;
    padding: 0;
}

.textarea { 
    @extend .text-input;
    height: 10rem;
}
