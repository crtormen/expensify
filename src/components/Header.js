import React from 'react';
import { Link, NavLink } from 'react-router-dom';
import { connect } from 'react-redux';
import { startLogout } from '../actions/auth';

export const Header = ({ startLogout }) => (
        <header>
            <h1>Expensify</h1>
            <div>
                <nav>
                <li><NavLink to="/dashboard" activeClassName="is-active" exact={true}>Home</NavLink></li>
                <li><NavLink to="/create" activeClassName="is-active">Add Expense</NavLink></li>
                <li><button onClick={startLogout}>Logout</button></li>
                </nav>
            </div>
        </header>
);

const mapDispatchToProps = (dispatch) => ({
    startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);