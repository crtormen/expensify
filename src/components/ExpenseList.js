import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses'

const onCheckboxChange = () => {
    console.log("CHECKBOX CHANGE");
};

export const ExpenseList = (props) => (
    <div>
    <h2>Expense List</h2>    
    {
        props.expenses.length === 0 ? (
            <p>No expenses</p>
        ) : (
            props.expenses.map((expense) => {
                return (
                        <ExpenseListItem
                            key={expense.id} 
                            {...expense} 
                        />
                )
            })
        )
    }
    </div>
);

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

export default connect(mapStateToProps)(ExpenseList); //connect to store to receive and show always the most updated list of expenses 

// const ConnectedExpenseList = connect((state) => {
//     return {
//         name: 'Claudio'
//     };

// })(ExpenseList);

// export default ConnectedExpenseList;