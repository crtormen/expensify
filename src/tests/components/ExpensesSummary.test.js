import React from 'react';
import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import { expenses } from '../fixtures/expenses';


test("should render ExpensesSummary with expenses", () => {
    const total = expenses.reduce((total, expense) => total + expense.amount);
    const wrapper = shallow(<ExpensesSummary expenseCount={expenses.length} expensesTotal={total} />);

    expect(wrapper).toMatchSnapshot();
});

test("should render ExpensesSummary with 1 expense", () => {
    const wrapper = shallow(<ExpensesSummary expenses={[expenses[0]]} />);

    expect(wrapper).toMatchSnapshot();    
})