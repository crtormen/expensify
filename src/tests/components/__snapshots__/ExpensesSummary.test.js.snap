// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpensesSummary with 1 expense 1`] = `
<div>
  <h3>
    Viewing 
     
    expenses
     totalling 
    $0.00
  </h3>
</div>
`;

exports[`should render ExpensesSummary with expenses 1`] = `
<div>
  <h3>
    Viewing 
    3
     
    expenses
     totalling 
    $0.00
  </h3>
</div>
`;
